crates/stackable-versioned-macros/src/lib.rs

```diff
@@ -8,6 +8,222 @@ mod attrs;
+///     version(name = "v1alpha1"),
+///     version(name = "v1beta1"),
+///     version(name = "v1"),
+///     version(name = "v2"),
+///     version(name = "v3")
+/// )]
+/// struct Foo {
+///     /// My docs
+///     #[versioned(
+///         added(since = "v1beta1"),
+///         renamed(since = "v1", from = "gau"),
+///         deprecated(since = "v2", note = "not empty")
+///     )]
+///     deprecated_bar: usize,
+///     baz: bool,
+/// }
+/// ```
+///
+/// ### Declaring Versions
+///
+/// Before any of the fields can be versioned, versions need to be declared at
+/// the container level. Each version currently supports two parameters: `name`
+/// and the `deprecated` flag. The `name` must be a valid (and supported)
+/// format. The macro checks each declared version and reports any error
+/// encountered during parsing.
+/// The `deprecated` flag marks the version as deprecated. This currently adds
+/// the `#[deprecated]` attribute to the appropriate piece of code.
+///
+/// ```
+/// # use stackable_versioned_macros::versioned;
+/// #[versioned(
+///     version(name = "v1alpha1", deprecated)
+/// )]
+/// struct Foo {}
+/// ```
+///
+/// Additionally, it is ensured that each version is unique. Declaring the same
+/// version multiple times will result in an error. Furthermore, declaring the
+/// versions out-of-order ist prohibited by default. It is possible to opt-out
+/// of this check by setting `options(allow_unsorted)`:
+///
+/// ```
+/// # use stackable_versioned_macros::versioned;
+/// #[versioned(
+///     version(name = "v1beta1"),
+///     version(name = "v1alpha1"),
+///     options(allow_unsorted)
+/// )]
+/// struct Foo {}
+/// ```
+///
+/// ### Field Actions
+///
+/// This library currently supports three different field actions. Fields can
+/// be added, renamed and deprecated. The macro ensures that these actions
+/// adhere to the following set of rules:
+///
+/// - Fields cannot be added and deprecated in the same version.
+/// - Fields cannot be added and renamed in the same version.
+/// - Fields cannot be renamed and deprecated in the same version.
+/// - Fields added in version _a_, renamed _0...n_ times in versions
+///   b<sub>1</sub>, b<sub>2</sub>, ..., b<sub>n</sub> and deprecated in
+///   version _c_ must ensure _a < b<sub>1</sub>, b<sub>2</sub>, ...,
+///   b<sub>n</sub> < c_.
+/// - All field actions must use previously declared versions. Using versions
+///   not present at the container level will result in an error.
+///
+/// ```
+/// # use stackable_versioned_macros::versioned;
+/// #[versioned(
+///     version(name = "v1alpha1"),
+///     version(name = "v1beta1"),
+///     version(name = "v1"),
+///     version(name = "v2"),
+/// )]
+/// struct Foo {
+///     #[versioned(
+///         added(since = "v1beta1"),
+///         renamed(since = "v1", from = "gau"),
+///         deprecated(since = "v2", note = "not empty")
+///     )]
+///     deprecated_bar: usize,
+///     baz: bool,
+/// }
```

Comment:
At a quick glance, this looks the same as the first code block.

Reply:
Removed in 3be5393